torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,395 @@
+      return it == preservedScalarAttrs_.end() || !it->second.count(name);
+    }
+
+    auto it = preservedAttrs_.find(attrModule._ivalue());
+    if (it == preservedAttrs_.end()) {
+      return true;
+    }
+    if (it->second.count(attr)) {
+      return false;
+    }
+    if (!attr.type()->cast<ClassType>()) {
+      for (auto& ivalue : it->second) {
+        if (!ivalue.isObject() && ivalue.overlaps(attr)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  void insertMutableAttr(
+      const std::string& name,
+      const IValue& attr,
+      script::Module& attrModule) {
+    if (AliasDb::mutableType(attr.type())) {
+      preservedAttrs_[attrModule._ivalue()].insert(attr);
+    } else {
+      preservedScalarAttrs_[attrModule._ivalue()].insert(name);
+    }
+  }
+
+  void recordMutableAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    std::unique_ptr<AliasDb> aliasDb =
+        torch::make_unique<AliasDb>(graph, /* isFrozen */ true);
+    IValue::HashAliasedIValues usedAttrs;
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr || n->kind() == prim::GetAttr) {
+          // TODO: handle interface attributes. For now, Exit if Module uses
+          // inteface attributes
+          if (n->kind() == prim::GetAttr) {
+            TORCH_CHECK(
+                !n->output()->type()->cast<InterfaceType>(),
+                "attempted to freeze a module that uses interface attributes");
+          }
+          auto inputNode = n->inputs()[0]->node();
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            continue;
+          }
+
+          auto attr = attrModule.attr(name);
+          if (n->kind() == prim::GetAttr) {
+            auto type = n->output()->type();
+            // Do not record submodules. Their attributes are tracked
+            // individually.
+            if (attr.isObject() || !AliasDb::mutableType(attr.type())) {
+              continue;
+            }
+            usedAttrs.insert(attr);
+          }
+
+          if (n->kind() == prim::SetAttr || aliasDb->hasOutputWriters(n)) {
+            GRAPH_DEBUG(
+                n->kind() == prim::GetAttr ? "attribute: " + name + " in %" +
+                        n->output()->debugName() + " has inplace writer"
+                                           : "attribute: " + name + " is set");
+            insertMutableAttr(name, attr, attrModule);
+          }
+        }
+      }
+    }
+    // FIXME: Current Alias analysis fails to track subvalues.
+    // This is not a common scenario, for freezing, detect and error out.
+    for (auto it = usedAttrs.begin(); it != usedAttrs.end();) {
```

Comment:
This is going to be quite slow because it's O(n^2) in # of usedAttrs, and `overlap` is O(n) in number of elements.
We could instantiate a new HashAliasedIValues, and then one by one iterate over usedAttrs, inserting all of it's subvalues into the new HashedAliasIValuess, checking if the element was already contained ? This will be O(n) instead of O(n^2).

This isn't strictly necessary, if there aren't many attributes this won't be that slow, but if we are recursing through other modules than it might become an issue.